device-connectors/src/testflinger_device_connectors/fw_devices/HPE/HPE.py

```diff
@@ -0,0 +1,461 @@
+"""Device class for flashing firmware on HPE server machines"""
+
+import subprocess
+import json
+import os
+import logging
+import time
+import requests
+import re
+from testflinger_device_connectors.fw_devices.base import OEMDevice
+from typing import Tuple
+
+logger = logging.getLogger()
+
+
+"""HPE firmware repository and index file"""
+HPE_SDR = "https://downloads.linux.hpe.com/SDR"
+HPE_SDR_REPO = f"{HPE_SDR}/repo"
+FW_REPOS = {
+    "rl": "rlcp",
+    "gen10": "fwpp-gen10",
+    "gen11": "fwpp-gen11",
+}
+INDEX_FILE = "fwrepo.json"
+
+
+class HPEDevice(OEMDevice):
+    """Device class for HPE server machines"""
+
+    fw_update_type = "OEM-defined"
+    vendor = ["HPE"]
+    reboot_timeout = 1800
+
+    def __init__(
+        self,
+        ipaddr: str,
+        user: str,
+        bmc_ip: str,
+        bmc_user: str,
+        bmc_password: str,
+    ):
+        super().__init__(ipaddr, user)
+        self.bmc_ip = bmc_ip
+        self.bmc_user = bmc_user
+        self.bmc_password = bmc_password
+        self.fwrepo_index = {}
+        self._install_ilorest()
+        self._login_ilo()
+
+    def _install_ilorest(self):
+        """Install stable/current ilorest deb from HPE SDR"""
+        install_cmd = [
+            f"curl -fsSL {HPE_SDR}/hpePublicKey2048_key1.pub"
+            " | sudo gpg --dearmor | sudo tee"
+            " /etc/apt/trusted.gpg.d/hpe.gpg > /dev/null",
+            f"echo '# HPE ilorest\n"
+            f"deb {HPE_SDR_REPO}/ilorest stable/current non-free'"
+            " > /etc/apt/sources.list.d/ilorest.list",
+            "apt update",
+            "apt install -y ilorest",
+        ]
+        for cmd in install_cmd:
+            subprocess.run(
+                cmd,
+                shell=True,
+                capture_output=True,
+            )
+        rc, stdout, stderr = self.run_cmd("--version")
+        if rc == 0:
+            logger.info("successfully installed %s" % stdout)
+        else:
+            raise RuntimeError("failed to install ilorest")
+
+    def run_cmd(
+        self, cmds: str, raise_stderr: bool = True, timeout: int = 30
+    ) -> Tuple[int, str, str]:
```

Comment:
If you’re using Python >=3.9, there is no need to use Tuple from typing library, you can just use `tuple` instead. Here are some examples in the doc: https://docs.python.org/3/library/typing.html#type-aliases